sycl/source/detail/sycl_mem_obj_t.hpp

```diff
@@ -276,6 +277,23 @@ class __SYCL_EXPORT SYCLMemObjT : public SYCLMemObjI {
 
   void markAsInternal() { MIsInternal = true; }
 
+  /// Returns true if this memory object requires a write_back on destruction.
+  bool needsWriteBack() const { return MNeedWriteBack && MUploadDataFunctor; }
+
+  /// Increment an internal counter for how many graphs are currently using this
+  /// memory object.
+  void markBeingUsedInGraph() { MGraphUseCount += 1; }
+
+  /// Decrement an internal counter for how many graphs are currently using this
+  /// memory object.
+  void markNoLongerBeingUsedInGraph() {
+    if (MGraphUseCount > 0)
+      MGraphUseCount -= 1;
```

Comment:
Don't think this comparison `(MGraphUseCount > 0)` then `-=` is thread safe, as it's two operations which don't make the whole function atomic

Reply:
Replaced this with a compare exchange loop so it is thread safe now.